### PR TITLE
DAOS-12734 common: Fix for failure/crash on container destroy

### PR DESCRIPTION
**Pull request by sherintg**
This fix addresses a bug within the bmem allocator, where the initial capacity for redo ops ulog was set incorrectly. Without this fix, if the number of redo log entries within a tx is between 40 and 64, then only the first 40 entries are getting committed.

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/common/dav/memops.c [hunk 1]
```diff
@@ -128,7 +128,7 @@ operation_log_persistent_init(struct operation_log *log,
 	}
 
 	/* initialize underlying redo log structure */
-	src->capacity = ulog_base_nbytes;
+	src->capacity = ULOG_BASE_SIZE;
```
Comment (NiuYawei) [inline src/common/dav/memops.c:131]: This looks really confusing to me. In dav_create_clogs(), we call ulog_construct_new() with LANE_REDO_EXTERNL_SIZE, but here we specify it as ULOG_BASE_SIZE?

It looks to me all the 'lane' stuff is unused? should we remove these LANE_REDO/UNDO size macros and always use ULOG_BASE_SIZE for creating & extending?

File: src/common/dav/memops.c [hunk 2]
```diff
@@ -531,7 +531,7 @@ operation_process_persistent_redo(struct operation_context *ctx)
 int
 operation_reserve(struct operation_context *ctx, size_t new_capacity)
 {
-	if (new_capacity > ctx->ulog_capacity) {
+	if ((ctx->type == LOG_TYPE_UNDO) && (new_capacity > ctx->ulog_capacity)) {
```
Comment (NiuYawei) [inline src/common/dav/memops.c:534]: I don't see why redo log doesn't need extending, could you explain it a bit?

File: src/common/dav/memops.c (hunk 1 above)
Comment (sherintg) [reply src/common/dav/memops.c:131]: The lane stuff in pmdk is to persist the redo and undo logs. The redo ops are initially logged into a transient memory logs (pshadow_ops and transient_ops) and later persisted to lane pointed by ctx->ulog before committing the changes. The undo logs are written directly to ulog as and when it happens.

With BMEM, there is no need to persist these logs and redo ops in pshadow_ops are converted to wal redo ops and not written to ctx->ulog. Hence the ulog_construct_new() with LANE_REDO_EXTERNL_SIZE can be cleaned up. The cleanup requires some more changes as certain functions expects ctx->ulog to be not NULL. @ivadmitry is looking at doing the cleanup with changes to the functions that operates on ctx->ulog.

File: src/common/dav/memops.c (hunk 2 above)
Comment (sherintg) [reply src/common/dav/memops.c:534]: As mentioned in the previous comment, the redo ops do not update ctx->ulog. Instead they get appended to the log within ctx->pshadow_ops.  The log in pshadow_ops is extended using realloc within the function operation_add_typed_entry().

**Review by NiuYawei — APPROVED** [review]
Good catch! and thanks for the explanation.

**Review by gnailzenh — APPROVED** [review]

**Pull request merged**